### DISCUSSION TITLE
who can tell me the source code where dotnet add package command run

### DISCUSSION THREAD

**Discussion by mingrenliu** [dotnet/runtime#91772] [Q&A] [answered]
i want to know how ''dotnet add package'' command change .csproj file. but when i read source code of dotnet/sdk , only find ForwardingApp class in namespace Microsoft.DotNet.Cli .the reference class ForwardingAppImplementation  not found. it maybe use local dotnet tool to run command,but i can not search deeply .who can tell me ,thanks.

**Comment by huoyaoyuan** [comment, marked as answer]
You can find the `ForwardingAppImplementation` type with GitHub symbol finder, or open the source code in Visual Studio:
![image](https://github.com/dotnet/runtime/assets/5644458/a5464c2d-2041-426e-a3e7-98384aa6df33)

https://github.com/dotnet/sdk/blob/a48ae5a2476abc529c6f4e9446384c4292c25611/src/Cli/dotnet/NuGetForwardingApp.cs#L10C48-L10C65
The package related commands are forwarded to NuGet. The actual working code is in [NuGet Client](https://github.com/NuGet/NuGet.Client) repo.

**Reply by mingrenliu** [reply]
thanks,i found it，nuget cli use Visual Studio Project System to change .csproj .

**Comment by mingrenliu** [comment]
> You can find the `ForwardingAppImplementation` type with GitHub symbol finder, or open the source code in Visual Studio: ![image](https://user-images.githubusercontent.com/5644458/266507124-a5464c2d-2041-426e-a3e7-98384aa6df33.png)
> 
> https://github.com/dotnet/sdk/blob/a48ae5a2476abc529c6f4e9446384c4292c25611/src/Cli/dotnet/NuGetForwardingApp.cs#L10C48-L10C65 The package related commands are forwarded to NuGet. The actual working code is in [NuGet Client](https://github.com/NuGet/NuGet.Client) repo.与包相关的命令将转发到 NuGet。实际工作代码位于 NuGet 客户端存储库中。
